### PR TITLE
Misc NDM Improvement

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
**Problem:**
1. Now we find out that NDM will remove the hot-plug device from the `blockdevices` that would cause the data to be lost when it add back.
2. Also, we allow the user to use the repair tool instead of the force format again.
3. Some improvement

**Solution:**
1. We changed the state to inactive for the hotplugged device instead of deleting it from `blockdevices` CR.
2. Add new field `corrupted`. For the `corrupted` device, we could choose the force format or try to use e2fsck to repair.
3. Log improvement
4. reduce the redundant CR update
5. remove the polling mechanism (But still keep two handling function now)

**Related Issue:**
https://github.com/harvester/harvester/issues/3218
https://github.com/harvester/harvester/issues/3412
https://github.com/harvester/harvester/issues/4016

**Test plan:**
case 1:
1. add a single disk
2. create external storage with the above disk
3. remove this disk
4. check the `blockdevices` status is inactive
5. add this disk back
6. check that the `blockdevices` status is active and should work again.

case 2:
**NOTE1: Please update the blockdevices CRD (on this PR) before we do this test**
**NOTE2: Please ensure you use the SCSI disk with wwn or the filesystem UUID will be erased when we try to break the filesystem.**
1. add a single disk
2. create external storage with the above disk
3. remove this storage
4. make this filesystem corrupted (you could use the following command `dd if=/dev/zero of=/dev/sda bs=512k count=1 oflag=sync`
5. try to create external storage with this disk
6. you should see the mount error and check the `Status.Filesystem.Corrupted` is true
7. Check on the UI that you could force format again.
8. Add it back with force format, and it should work again.

### CODE REVIEW

File: pkg/controller/blockdevice/controller.go [hunk 1]
```diff
@@ -155,14 +155,9 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
 			diskv1.DeviceFormatting.SetError(deviceCpy, "", err)
 			diskv1.DeviceFormatting.SetStatusBool(deviceCpy, false)
 		}
-		if !reflect.DeepEqual(device, deviceCpy) {
-			logrus.Debugf("Update block device %s for new formatting state", device.Name)
-			return c.Blockdevices.Update(deviceCpy)
-		}
-		return device, err
 	}
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:158]: removing the crd update, is likely to trigger reformatting if the mount fails for any reason. would this be expected behaviour?

File: pkg/controller/blockdevice/scanner.go [hunk 2]
```diff
@@ -216,6 +234,11 @@ func (s *Scanner) SaveBlockDevice(bd *diskv1.BlockDevice, autoProvisioned bool)
 		bd.Spec.FileSystem.ForceFormatted = true
 		bd.Spec.FileSystem.Provisioned = true
 	}
+	if curBd, err := s.Blockdevices.Get(bd.Namespace, bd.Name, metav1.GetOptions{}); err == nil {
+		logrus.Infof("The inactive block device %s is coming back", bd.Name)
+		curBd.Status.State = diskv1.BlockDeviceActive
+		return s.Blockdevices.Update(curBd)
+	}
 	logrus.Infof("Add new block device %s with device: %s", bd.Name, bd.Spec.DevPath)
 	return s.Blockdevices.Create(bd)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/scanner.go:243]: we are not handling the api call errors. In such a case the code will try and create the object, when it may still already exist and fail.

we could may be re-organize this as 

```go
	curBd, err := s.Blockdevices.Get(bd.Namespace, bd.Name, metav1.GetOptions{})
	if err == nil {
		if apierrors.IsNotFound(err) {
			if autoProvisioned {
				bd.Spec.FileSystem.ForceFormatted = true
				bd.Spec.FileSystem.Provisioned = true
			}
			logrus.Infof("Add new block device %s with device: %s", bd.Name, bd.Spec.DevPath)
			return s.BlockDevices.Create(bd)
		}
		return err
	}
	logrus.Infof("The inactive block device %s is coming back", bd.Name)
	curBd.Status.State = diskv1.BlockDeviceActive
	return s.Blockdevices.Update(curBd)
```

File: pkg/controller/blockdevice/controller.go [hunk 3]
```diff
@@ -243,24 +238,30 @@ func (c *Controller) updateDeviceMount(device *diskv1.BlockDevice, devPath strin
...
 		logrus.Infof("Mount deivce %s to %s", device.Name, expectedMountPoint)
-		if err := disk.MountDisk(devPath, expectedMountPoint); err != nil {
+		if err := util.MountDisk(devPath, expectedMountPoint); err != nil {
+			if util.IsFSCorrupted(err) {
+				logrus.Errorf("Target device may be corrupted, update FS info.")
+				devCopy := device.DeepCopy()
+				devCopy.Status.DeviceStatus.FileSystem.Corrupted = true
+				if !reflect.DeepEqual(device, devCopy) {
+					logrus.Debugf("Update block device %s for new DeviceStatus state", device.Name)
+					c.Blockdevices.Update(devCopy)
```
Comment (chrisho) [inline pkg/controller/blockdevice/controller.go:248]: missing handle the error return

File: pkg/controller/blockdevice/scanner.go [hunk 4]
```diff
@@ -154,9 +160,19 @@ func (s *Scanner) scanBlockDevicesOnNode() error {
 	// This oldBds are leftover after running SaveBlockDevice.
 	// Clean up all previous registered block devices.
```
Comment (chrisho) [inline pkg/controller/blockdevice/scanner.go:161]: The comment should be updated to describe the following code

Comment (chrisho) [inline pkg/controller/blockdevice/scanner.go:161]: The comment should be updated to describe the following code

File: pkg/controller/blockdevice/controller.go [hunk 5]
```diff
@@ -243,24 +238,30 @@ func (c *Controller) updateDeviceMount(device *diskv1.BlockDevice, devPath strin
...
+				}
+			}
 			return err
```
Comment (chrisho) [inline pkg/controller/blockdevice/controller.go:251]: It seems that after updating here and returning an error, the caller may fail to update since it has changed here and the old obj is outdated.

File: pkg/controller/blockdevice/scanner.go (hunk 2 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go]: nice tips, update!

File: pkg/controller/blockdevice/controller.go (hunk 3 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/controller.go]: Thanks for catching this! fixed

File: pkg/controller/blockdevice/scanner.go (hunk 4 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go]: OK, updated!

File: pkg/controller/blockdevice/controller.go (hunk 5 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/controller.go:257]: Yap, I miss this situation. Thanks for it. Fixed!

**Comment by chrisho** [comment]
Could you supplement the test plan?

**Comment by Vicente-Cheng** [comment]
> Could you supplement the test plan?

Sorry, miss for it. Updated!

File: deploy/charts/harvester-node-disk-manager/templates/daemonset.yaml [hunk 6]
```diff
@@ -30,6 +30,7 @@ spec:
         imagePullPolicy: {{ .Values.image.pullPolicy }}
         command:
         - node-disk-manager
+        - --debug
```
[the remark this answers is not in the document]
Comment (Vicente-Cheng) [reply deploy/charts/harvester-node-disk-manager/templates/daemonset.yaml]: fixed!

File: pkg/apis/harvesterhci.io/v1beta1/types.go [hunk 7]
```diff
@@ -159,6 +162,9 @@ type FilesystemStatus struct {
 
 	// the last force formatted timestamp, only exist when user operate device formatting through the CRD controller
 	LastFormattedAt *metav1.Time `json:"LastFormattedAt,omitempty"`
+
+	// idicating whether the filesystem is corrupted or not
```
[the remark this answers is not in the document]
Comment (Vicente-Cheng) [reply pkg/apis/harvesterhci.io/v1beta1/types.go]: fixed!

File: pkg/udev/uevent.go [hunk 8]
```diff
@@ -109,11 +109,19 @@ func (u *Udev) ActionHandler(uevent netlink.UEvent) {
 			logrus.Infof("Skip adding non-identifiable block device %s", bd.Spec.DevPath)
 			return
 		}
+		u.scanner.Cond.L.Lock()
 		autoProvisioned := udevDevice.IsDisk() && u.scanner.ApplyAutoProvisionFiltersForDisk(disk)
 		u.AddBlockDevice(bd, autoProvisioned)
+		logrus.Infof("Wake up scanner with %s operation", netlink.ADD)
```
Comment (chrisho) [inline pkg/udev/uevent.go:115]: Add DevName to info message.

File: pkg/udev/uevent.go [hunk 9]
```diff
@@ -109,11 +109,19 @@ func (u *Udev) ActionHandler(uevent netlink.UEvent) {
...
+		u.scanner.Cond.Signal()
+		u.scanner.Cond.L.Unlock()
 	case netlink.REMOVE:
 		if udevDevice.IsDisk() {
-			u.RemoveBlockDevice(bd, &udevDevice, disk)
+			// just wake up scanner to check if the disk is removed, do no-op internally
+			u.scanner.Cond.L.Lock()
+			logrus.Infof("Wake up scanner with %s operation", netlink.REMOVE)
```
Comment (chrisho) [inline pkg/udev/uevent.go:122]: ditto

File: pkg/util/util.go [hunk 10]
```diff
@@ -49,3 +62,113 @@ func ContainsIgnoredCase(s []string, k string) bool {
...
+func MakeExt4DiskFormatting(devPath, uuid string) error {
+	args := []string{"-F", devPath}
+	if uuid != "" {
+		args = append(args, "-U", uuid)
+	}
+	cmd := exec.Command("mkfs.ext4", args...)
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
```
Comment (chrisho) [inline pkg/util/util.go:75]: Prefer use `cmd.CombinedOutput()` instead.
```go
if _, err := cmd.CombinedOutput(); err != nil {
  return fmt.Errorf("stderr: %s, err: %s", stderr.String(), err.Error())
}
```

File: main.go [hunk 11]
```diff
@@ -250,5 +256,11 @@ func run(opt *option.Option) error {
 	start(ctx)
 
 	<-ctx.Done()
+	scanner.Shutdown = true
+	logrus.Infof("NDM is going shutdown")
+	scanner.Cond.L.Lock()
+	scanner.Cond.Signal()
+	scanner.Cond.L.Unlock()
+	blockdevicev1.WaitGroup.Wait()
```
Comment (chrisho) [inline main.go:264]: More grace to use block channel to instead

**Comment by chrisho** [comment]
Could you create more detailed info to the wiki page? e.g. SCSI disk dependency and WWN.

File: pkg/controller/blockdevice/controller.go [hunk 12]
```diff
@@ -129,12 +127,19 @@ func Register(
...
+		return nil, nil
+	}
+
+	// corrupted device could be skipped if we do not set ForceFormatted or Repaired
+	if device.Status.DeviceStatus.FileSystem.Corrupted && !device.Spec.FileSystem.ForceFormatted && !device.Spec.FileSystem.Repaired {
 		return nil, nil
 	}
 
 	deviceCpy := device.DeepCopy()
+	logrus.Infof("[DEBUG] device: %+v", device)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:141]: are we happy leaving this logging statement at Info level?

File: pkg/controller/blockdevice/controller.go [hunk 13]
```diff
@@ -161,8 +168,9 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
 		}
 		return device, err
 	}
+	logrus.Infof("[DEBUG] #1 device.Status.ProvisionPhase: %v", device.Status.ProvisionPhase)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:171]: please review logging level

File: pkg/controller/blockdevice/controller.go [hunk 14]
```diff
@@ -171,11 +179,12 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
 			diskv1.DeviceMounted.SetStatusBool(deviceCpy, false)
 		}
 		if !reflect.DeepEqual(device, deviceCpy) {
-			logrus.Debugf("Update block device %s for new mount state", device.Name)
+			logrus.Debugf("Update block device %s for new formatting and mount state", device.Name)
 			return c.Blockdevices.Update(deviceCpy)
 		}
 		return device, err
 	}
+	logrus.Infof("[DEBUG] #2 device.Status.ProvisionPhase: %v", device.Status.ProvisionPhase)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:187]: please review logging level

File: pkg/controller/blockdevice/controller.go [hunk 15]
```diff
@@ -188,6 +197,7 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
 	 *       check with `device.Status.ProvisionPhase`
 	 */
 	needProvision := deviceCpy.Spec.FileSystem.Provisioned
+	logrus.Infof("[DEBUG] needProvision: %v, device.Status.ProvisionPhase: %v", needProvision, device.Status.ProvisionPhase)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:200]: please review logging level

File: pkg/controller/blockdevice/controller.go [hunk 16]
```diff
@@ -211,7 +221,18 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
 
 	if !reflect.DeepEqual(device, deviceCpy) {
 		logrus.Debugf("Update block device %s for new provision state", device.Name)
-		return c.Blockdevices.Update(deviceCpy)
+		logrus.Infof("[DEBUG] update the device status: %v", deviceCpy.Status.ProvisionPhase)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:224]: please review log level

File: pkg/controller/blockdevice/controller.go [hunk 17]
```diff
@@ -211,7 +221,18 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
...
+		dev, err := c.Blockdevices.Update(deviceCpy)
+		if err != nil {
+			logrus.Errorf("failed to update block device %s: %v", device.Name, err)
+			diskv1.DiskAddedToNode.SetError(deviceCpy, "", err)
+			diskv1.DiskAddedToNode.SetStatusBool(deviceCpy, false)
+			c.Blockdevices.EnqueueAfter(c.Namespace, device.Name, jitterEnqueueDelay())
+			return nil, err
+		}
+		logrus.Infof("[DEBUG] updated device: %+v", dev)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:233]: please review log level

File: pkg/controller/blockdevice/controller.go [hunk 18]
```diff
@@ -243,24 +266,31 @@ func (c *Controller) updateDeviceMount(device *diskv1.BlockDevice, devPath strin
...
 	}
-	if needMountUpdate != NeedMountUpdateNo {
-		return c.updateDeviceFileSystem(device, devPath)
-	}
-	return nil
+	device.Status.DeviceStatus.FileSystem.Corrupted = false
+	return c.updateDeviceFileSystem(device, devPath)
 }
 
 func (c *Controller) updateDeviceFileSystem(device *diskv1.BlockDevice, devPath string) error {
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:284]: please review log level

File: pkg/controller/blockdevice/controller.go [hunk 19]
```diff
@@ -354,17 +385,21 @@ func (c *Controller) provisionDeviceToNode(device *diskv1.BlockDevice) error {
 	if err != nil {
 		return err
 	}
+	logrus.Infof("[DEBUG] Get latest disks on node %s, %+v", c.NodeName, node.Spec.Disks)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:388]: please review log level

File: pkg/controller/blockdevice/controller.go [hunk 20]
```diff
@@ -354,17 +385,21 @@ func (c *Controller) provisionDeviceToNode(device *diskv1.BlockDevice) error {
...
+	diskSpec := longhornv1.DiskSpec{
 		Path:              extraDiskMountPoint(device),
 		AllowScheduling:   true,
 		EvictionRequested: false,
 		StorageReserved:   0,
 		Tags:              []string{},
 	}
 
-	if disk, ok := node.Spec.Disks[device.Name]; !ok || !reflect.DeepEqual(disk, diskSpec) {
+	logrus.Infof("[DEBUG] checked diskSpec %+v", diskSpec)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:399]: please review log level

File: pkg/controller/blockdevice/controller.go [hunk 21]
```diff
@@ -374,6 +409,7 @@ func (c *Controller) provisionDeviceToNode(device *diskv1.BlockDevice) error {
 		}
 
 		if !diskv1.DiskAddedToNode.IsTrue(device) {
+			logrus.Infof("[DEBUG] Update to ProvisionPhaseProvisioned and set DiskAddedToNode to true for device %s", device.Name)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:412]: please review log level

File: pkg/controller/blockdevice/controller.go [hunk 22]
```diff
@@ -459,6 +496,9 @@ func (c *Controller) updateDeviceStatus(device *diskv1.BlockDevice, devPath stri
 	var newStatus diskv1.DeviceStatus
 	var needAutoProvision bool
 
+	if diskv1.DiskAddedToNode.IsTrue(device) {
+		logrus.Infof("[DEBUG] we do not update device from OS side if DiskAddedToNode is true for device %s", device.Name)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/controller.go:500]: please review log level

File: pkg/controller/blockdevice/scanner.go [hunk 23]
```diff
@@ -151,12 +162,22 @@ func (s *Scanner) scanBlockDevicesOnNode() error {
...
+			logrus.Debugf("The device %s is already inactive, continue.", oldBd.Name)
+			continue
+		}
+		logrus.Debugf("Change the device %s to inactive.", oldBd.Name)
+		newBd := oldBd.DeepCopy()
+		newBd.Status.State = diskv1.BlockDeviceInactive
+		if !reflect.DeepEqual(oldBd, newBd) {
+			logrus.Debugf("Update block device %s for new formatting and mount state", oldBd.Name)
+			if _, err := s.Blockdevices.Update(newBd); err != nil {
+				logrus.Fatalf("Update device %s status error", oldBd.Name)
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/scanner.go:178]: Fatal will cause the controller process to be killed? Would this be the expected behaviour?

Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go]: Nice catch! Change to `Errorf`.

File: pkg/util/util.go (hunk 10 above)
Comment (Vicente-Cheng) [reply pkg/util/util.go]: good suggestion!  updated!

File: main.go (hunk 11 above)
Comment (Vicente-Cheng) [reply main.go]: Sure, Change to use block channel for stopping

**Comment by Vicente-Cheng** [comment]
> Could you create more detailed info to the wiki page? e.g. SCSI disk dependency and WWN.

Added with following commits
https://github.com/harvester/node-disk-manager/pull/55/commits/f8a49911710a4c5dc6fa59b928d12575bb5d2426

File: pkg/controller/blockdevice/scanner.go [hunk 24]
```diff
@@ -16,9 +16,7 @@ import (
 	ctldiskv1 "github.com/harvester/node-disk-manager/pkg/generated/controllers/harvesterhci.io/v1beta1"
 )
 
-const (
-	defaultRescanInterval = 30 * time.Second
-)
+var WaitGroup sync.WaitGroup
```
Comment (chrisho) [inline pkg/controller/blockdevice/scanner.go:19]: remove if unused

File: main.go [hunk 25]
```diff
@@ -179,6 +184,7 @@ func initLogs(opt *option.Option) {
 
 func run(opt *option.Option) error {
 	logrus.Info("Starting node disk manager controller")
+	blockdevicev1.WaitGroup.Add(1)
```
Comment (chrisho) [inline main.go:187]: remove if unused

File: pkg/controller/blockdevice/scanner.go (hunk 24 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go]: Thanks for the reminder! updated!

File: main.go (hunk 25 above)
Comment (Vicente-Cheng) [reply main.go]: updated!

**Review by chrisho — APPROVED** [review]
LGTM, cases removed disk can add it back, and the corrupted disk can be force format and work fine after, thanks.

File: pkg/controller/blockdevice/scanner.go [hunk 26]
```diff
@@ -216,8 +233,21 @@ func (s *Scanner) SaveBlockDevice(bd *diskv1.BlockDevice, autoProvisioned bool)
...
 	}
-	logrus.Infof("Add new block device %s with device: %s", bd.Name, bd.Spec.DevPath)
-	return s.Blockdevices.Create(bd)
+	curBd, err := s.Blockdevices.Get(bd.Namespace, bd.Name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if autoProvisioned {
+				bd.Spec.FileSystem.ForceFormatted = true
+				bd.Spec.FileSystem.Provisioned = true
+			}
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/scanner.go:242]: not needed, as this is already checked and setup in the outer block

File: pkg/controller/blockdevice/scanner.go [hunk 27]
```diff
@@ -216,8 +233,21 @@ func (s *Scanner) SaveBlockDevice(bd *diskv1.BlockDevice, autoProvisioned bool)
...
+			logrus.Infof("Add new block device %s with device: %s", bd.Name, bd.Spec.DevPath)
+			return s.Blockdevices.Create(bd)
+		}
+		return nil, err
+	}
+	logrus.Infof("The inactive block device %s is coming back", bd.Name)
+	curBd.Status.State = diskv1.BlockDeviceActive
```
Comment (ibrokethecloud) [inline pkg/controller/blockdevice/scanner.go:249]: this may be an edge case.. but may happen in real life. remove a disk and add a new disk. the disk slot identifier is the same... say /dev/sdc, however this is a new disk, and we are just marking the disk as active.

in my tests on libvirt, i added a disk /dev/sdc, with a wwn, which was duely processed and added to a blockdevice.

I removed this disk, and deleted the disk, and added a new disk, which would be similar to replacing a disk in the same host slot.

the new disk gets the same device identifier /dev/sdc, but the underlying disk is different, and has a different wwn, but is processed as an updated and device status does not match the physical disk.

should be add more checks here like 
```suggestion
    if curBd.Status.State.Details.WWN != bd.Status.State.Details.WWN {
        // delete and create a new BD so it can fall through the new block device processing loop
    }
```

Comment (ibrokethecloud) [reply pkg/controller/blockdevice/scanner.go:249]: closing this.. could not re-produce this issue.

Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go:245]: Thanks @ibrokethecloud for taking care of this!
I add some logs to record the `wwn` if we meet this again.

File: pkg/controller/blockdevice/scanner.go (hunk 26 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go:238]: Updated! Remove the logic on the outer block because we only create bd with ENOENT.

**Review by ibrokethecloud — APPROVED** [review]
LGTM. thanks a lot for your help.

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

File: pkg/controller/blockdevice/controller.go [hunk 28]
```diff
@@ -228,9 +228,10 @@ func (c *Controller) OnBlockDeviceChange(key string, device *diskv1.BlockDevice)
 	return nil, nil
 }
 
-func (c *Controller) updateDeviceMount(device *diskv1.BlockDevice, devPath string, filesystem *block.FileSystemInfo, needMountUpdate NeedMountUpdate) error {
+func (c *Controller) updateDeviceMount(device *diskv1.BlockDevice, devPath string, filesystem *block.FileSystemInfo, needMountUpdate NeedMountUpdateOP) error {
+	logrus.Infof("Prepare to try %s", convertMountStr(needMountUpdate))
 	if device.Status.DeviceStatus.Partitioned {
-		return fmt.Errorf("cannot mount device with partitions")
+		return fmt.Errorf("not support the partitioned device now")
```
[the remark this answers is not in the document]
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/controller.go]: fixed!

File: main.go [hunk 29]
```diff
@@ -250,5 +256,11 @@ func run(opt *option.Option) error {
 	start(ctx)
 
 	<-ctx.Done()
+	scanner.Shutdown = true
+	logrus.Infof("NDM is going shutdown")
```
[the remark this answers is not in the document]
Comment (Vicente-Cheng) [reply main.go]: fixed!

**Comment by bk201** [comment]
Try to attach some disks in a very short time, but only some are created:
```
for ((i=0;i<10;i++)); do ./attach-disk.sh node2 ; done
```

Block devices (please ignore node3)
```
NAMESPACE         NAME                               TYPE   DEVPATH    MOUNTPOINT   NODENAME   STATUS   AGE
longhorn-system   5819b466585e1e3b1c9ed398918a9cb1   disk   /dev/sdi                node2      Active   5m4s
longhorn-system   8a655634522b710cc4f2d3514f10b5c3   disk   /dev/sdc                node3      Active   10m
longhorn-system   91c3333ea91016f699f1641ba3228354   disk   /dev/sda                node2      Active   5m11s
longhorn-system   a7f262a2e9610974dd9904ee577f7515   disk   /dev/sdf                node2      Active   5m8s
longhorn-system   b3d51f437d5c71e7476929625234549e   disk   /dev/sdd                node3      Active   10m
longhorn-system   d390352fbf27678e80eed80342918336   disk   /dev/sde                node3      Active   10m
```

Virtual disks:
```
node2:~ # ls -alh /dev/sd*
brw-rw---- 1 root disk 8,   0 Jul  3 06:38 /dev/sda
brw-rw---- 1 root disk 8,  16 Jul  3 06:24 /dev/sdb
brw-rw---- 1 root disk 8,  32 Jul  3 06:38 /dev/sdc
brw-rw---- 1 root disk 8,  48 Jul  3 06:38 /dev/sdd
brw-rw---- 1 root disk 8,  64 Jul  3 06:38 /dev/sde
brw-rw---- 1 root disk 8,  80 Jul  3 06:38 /dev/sdf
brw-rw---- 1 root disk 8,  96 Jul  3 06:38 /dev/sdg
brw-rw---- 1 root disk 8, 112 Jul  3 06:38 /dev/sdh
brw-rw---- 1 root disk 8, 128 Jul  3 06:38 /dev/sdi
brw-rw---- 1 root disk 8, 144 Jul  3 06:38 /dev/sdj
brw-rw---- 1 root disk 8, 160 Jul  3 06:38 /dev/sdk
```

NDM log:
[harvester-node-disk-manager-4v85m.log](https://github.com/harvester/node-disk-manager/files/11933894/harvester-node-disk-manager-4v85m.log)

File: pkg/apis/harvesterhci.io/v1beta1/types.go [hunk 30]
```diff
@@ -16,7 +16,7 @@ var (
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 // +kubebuilder:resource:shortName=bd;bds,scope=Namespaced
 // +kubebuilder:printcolumn:name="Type",type="string",JSONPath=`.status.deviceStatus.details.deviceType`
-// +kubebuilder:printcolumn:name="DevPath",type="string",JSONPath=`.spec.devPath`
+// +kubebuilder:printcolumn:name="DevPath",type="string",JSONPath=`.status.devStatus.devPath`
```
Comment (bk201) [inline pkg/apis/harvesterhci.io/v1beta1/types.go:19]:
```suggestion
// +kubebuilder:printcolumn:name="DevPath",type="string",JSONPath=`.status.deviceStatus.devPath`
```

Comment (Vicente-Cheng) [reply pkg/apis/harvesterhci.io/v1beta1/types.go]: nice catch! fixed!
